README.md
## Quick guide: build and run IoTivity projects on Ubuntu

 1. Build
	
	  Go to the top directory of `iotivity` project (Note: should always run `scons`
    command in this directory)

    Install external libraries:
    
        $ sudo apt-get install libboost-dev libboost-program-options-dev libexpat1-dev libboost-thread-dev

    Build release binaries:
    
        $ scons

    (Note: C++ sdk requires cereal. Please follow the instruction in the build
    message to install cereal)

    Build debug binaries:
    
        $scons RELEASE=false

    Help:
    
        $ scons -h

    Clear:
    
        $ scons -c

 2. Run the samples
 
        $ export LD_LIBRARY_PATH=<iotivity>/out/linux/x86_64/release
      
    Run the c++ samples in `<iotivity>/out/linux/x86_64/release/resource/examples`
      
    Run the c samples in `<iotivity>/out/linux/x86_64/release/resource/csdk/stack/samples/linux/SimpleClientServer`
      
    (`<iotivity>` is the path to `iotivity` project. If your device is x86, arm,
    or arm64, please change `x86_64` to the proper arch)

## How to build IoTivity projects

IoTivity includes a series of projects. You can find all these projects here:
    https://gerrit.iotivity.org/gerrit/#/admin/projects/

You can build IoTivity project on Linux / Windows / MAC OSX for various OS 
(Linux, Tizen, Android, Arduino, Windows, MAC OSX, IOS ...).

The output of the build is in:
  `<top directory of the project>/out/<target_os>/<target_arch>/<build_type>/`
e.g.
	`oic-resource/out/android/armeabi-v7a/release/`.

This document takes `iotivity` project as example, the way to build other
projects is almost the same.

### IoTivity project build tool scons

Scons is a cross-platform build tool, its usage is quite similar to GNU make.

To build a project, you just require to run following command at the directory
where a SConstruct file exists (SConstruct is the entrance of scons build, it's
equivalent to Makefile of `make`):

    $ scons [options] [target]

In additional, usually the scons build script of a project provides useful help
information(include build options). To see the help information:

    $ scons [options] -h

Note: If no value is specified for an option, the default value will be used.
The change of options value may impact the output.

### Prerequites

1. Scons

  Please refer to the following page to install scons:
   http://www.scons.org/doc/production/HTML/scons-user.html#chap-build-install
  (Note: on Windows, install Python 2.x before installing scons)

2. IDE/SDK Prerequites

  To build for some OS (Android / Arduino / IOS ...), an IDE/SDK may be required,
  please go to the relative page to download and install the required IDE/SDK.

  Android:

  To build for Android, Andorid NDK and SDK are required.
  
  Android NDK: http://developer.android.com/tools/sdk/ndk/index.html
  
  Android SDK: http://developer.android.com/sdk/index.html
  
  (Note: as in some IoTivity projects, C++11 features are used, recommend Android
  NDK >= r10, according to our test result r10c is the best one currently)

  Arduino:

  To build for Arduino, Arduino IDE is required.
  
  Arduino IDE: http://arduino.cc/en/Main/Software
  
  (Note: recommend install Arduino IDE >=1.5.7)

  Arduino builds are dependent on latest Time library. Download it from here:
    http://www.pjrc.com/teensy/td_libs_Time.html
  and extract to `<arduino_ide_root>/libraries/`

  (Note: If you are using Arduino IDE 1.5.8 BETA on Windows, it may pop up some
  dll isn't found. please copy relative dll from the IDE directory to
  C:\Windows\SysWOW64. IDE 1.5.7 doesn't have this issue. Other version IDE isn't
  tested.)

  Apple:
  
  To build for Mac OSX or IOS, Xcode is required.
  
  Xcode: https://developer.apple.com/xcode/downloads/

  Java:
  
  To build the Java code, JDK is required.
  
  JDK: http://www.oracle.com/technetwork/java/javase/downloads/index.html
  
  (If the project doesn't include Java code or you wouldn't like build the
  Java codes, this isn't required)

  (Note: for convenience, suggest add the IDE/SDK path in environment variable,
  so you don't need to add it in command line each time. The build script will
  guide you to do that.)

3. External libraries

  For Android and IOS build, most of the external libraries are provided as
  binary in oic-utilities project (https://oic-review.01.org/gerrit/oic-utilities).
  Please download it in the same directory as other IoTivity projects. If it's
  in different directory, an additional option (OIC_UITLS) will be required. The
  build command should be:
  
      $ scons OIC_UITLS=<path to oic-utilities> [other options] [target]

  (Note: for convenience, you can also add it in environment variable. So you
  don't need to add this option in command line each time.)

  e.g.:
  
      $ export OIC_UITLS=<path to oic-utilities project>

### Build IoTivity project on Linux (Ubuntu) ###

Generally, it's required to specify the target OS and target ARCH, that's to say
tell Scons which OS and which ARCH you'd like build this project for. By default,
the target OS and ARCH is the same as the host.

Some more options may be required, please care the `error` notification when build.
For help about how to set an option, please run:

	$ scons TARGET_OS=xxx TARGET_ARCH=yyy [XXX=zzz ...] -h

1. Build IoTivity project for Linux

        $ cd <top directory of the project>
        $ sudo apt-get install libboost-dev libboost-program-options-dev libexpat1-dev libboost-thread-dev
        $ scons

2. Build IoTivity project for Android

        $ cd <top directory of the project>
        $ scons TARGET_OS=android TARGET_ARCH=xxx
      
  (xxx can be x86, armeabi, armeabi-v7a, armeabi-v7a-hard. To see all of its
  allowed value, please execute command `scons TARGET_OS=android -Q -h`)
  
  Note: Currently as x86_64/arm64_v8a external library binaries aren't provided,
  you may meet link problem if build executable binary which depends on external
  library for x86_64/arm64_v8a.

3. Build IoTivity project for Arduino

        $ cd <top directory of the project>
        $ scons TARGET_OS=arduino TARGET_ARCH=xxx BOARD=yyy
      
  (xxx can be avr, arm; yyy is the name of the board, to get its allowed value
  run: scons TARGET_OS=arduino TARGET_ARCH=xxx -h. You may see a option `CPU` in
  the output of above command line, that's due to some boards have different
  processors, to specify the processor, add `CPU=zzz` in the command line. If no
  `CPU` option exists, that means the board only support one kind of processor,
  it's unnecessary to specify it)

4. Build Iotivity project for Tizen

        $ cd <top directory of the project>
        $ gbs build -A xxx --packaging-dir tools/tizen/
      
  (xxx can be i586 or armv7l, we provide the spec file required by gbs tool at
  toools/tizen directory. gbs is default build tool for Tizen platfrom, we can
  refer the following wiki to setup Tizen development environment:
  https://source.tizen.org/documentation/developer-guide/getting-started-guide)

### Build IoTivity project on Windows

1. Build IoTivity project for Android (It's the same as on Ubuntu)

        $ cd <top directory of the project>
        $ scons TARGET_OS=android TARGET_ARCH=xxx
      
  (xxx can be x86, armeabi, armeabi-v7a, armeabi-v7a-hard ...)

2. Build IoTivity project for Arduino (It's the same as on Ubuntu)

        $ cd <top directory of the project>
        $ scons TARGET_OS=arduino TARGET_ARCH=xxx BOARD=yyy
      
  (xxx can be avr, arm; yyy is the name of the board, to get its allowed value
  run: scons TARGET_OS=arduino TARGET_ARCH=xxx -h. You may see a option `CPU` in
  the output of above command line, that's due to some boards have different
  processor, to specify the processor, add `CPU=zzz` in the command line. If no
  `CPU` option exists, that means the board only support one kind of processor,
  it's unnecessary to specify it)

  Note: Currently most IoTivity project doesn't support Windows, so you can't set
  TARGET_OS to `windows` except the project support Windows.

  That's to say if the project doesn't support Windows, run:
  
      $ scons TARGET_OS=windows ....
        
  or run on Windows
  
      $ scons
        
  may always fail.

### Build IoTivity project on Mac OSX

1. Build IoTivity project for Mac OSX

        $ cd <top directory of the project>
        $ scons SYS_VERSION=yyy
      
  (yyy is the OSX version, e.g. 10.9)

2. Build IoTivity project for Android (It's the same as on Ubuntu)

        $ cd <top directory of the project>
        $ scons TARGET_OS=android TARGET_ARCH=xxx
      
  (xxx can be x86, armeabi, armeabi-v7a, armeabi-v7a-hard)

3. Build IoTivity project for IOS

        $ cd <top directory of the project>
        $ scons TARGET_OS=ios TARGET_ARCH=xxx SYS_VERSION=yyy
      
  (xxx can be i386, x86_64, armv7, armv7s, arm64, yyy is IOS version, e.g. 7.0)
  
  Note:
  
  1) for convenience, a script (auto_build.sh) is provided to run possible build
  at once. Following is the usage:
  
  To build:
  
      $ auto_build.sh <path-to-android-ndk> <path-to-arduino-home>
       
  To clean:
  
      $ auto_build.sh -c
  
  2) For Arduino build, the Time library should >=1.3. The old can only be built
  with Arduino IDE 1.0.x


debian/changelog
iotivity (0.0.0-0) unstable; urgency=low

  * Initial release (Closes: #824155)

 -- Philippe Coval <rzr@gna.org>  Tue, 01 Dec 2015 23:32:27 +0000


auto_build.sh
#! /bin/bash

# Ideally we will capture the exit code of each step and try them all before failing
# the build script.  For now, use set -e and fail the build at first failure.
set -e

function build_all()
{
	if [ $(uname -s) = "Linux" ]
        then
		build_linux_unsecured $1 $2
		build_linux_secured $1 $2
		build_linux_unsecured_with_ra $1 $2
		build_linux_secured_with_ra $1 $2
		build_linux_unsecured_with_rm $1 $2
		build_linux_unsecured_with_rd $1 $2
		build_linux_secured_with_rd $1 $2
		build_simulator $1 $2
	fi

	build_android $1 $2

	build_arduino $1 $2

	build_tizen $1 $2

	if [ $(uname -s) = "Darwin" ]
	then
		build_darwin $1 $2
	fi
}

function build_linux()
{
	build_linux_unsecured $1 $2

	build_linux_secured $1 $2
}

function build_linux_unsecured()
{
	echo "*********** Build for linux ************"
	scons RELEASE=$1 $2
}

function build_linux_unsecured_with_rm()
{
	echo "*********** Build for linux with RoutingManager************"
	scons ROUTING=GW RELEASE=$1 $2
}

function build_linux_secured()
{
	echo "*********** Build for linux with Security *************"
	scons RELEASE=$1 SECURED=1 $2
}

function build_linux_unsecured_with_ra()
{

	echo "*********** Build for linux With Remote Access *************"
	scons RELEASE=$1 WITH_RA=1 WITH_RA_IBB=1 $2
}

function build_linux_secured_with_ra()
{
	echo "*********** Build for linux With Remote Access & Security ************"
	scons RELEASE=$1 WITH_RA=1 WITH_RA_IBB=1 SECURED=1 $2
}

function build_linux_unsecured_with_rd()
{
	echo "*********** Build for linux With Resource Directory *************"
	scons RELEASE=$1 WITH_RD=1 $2
}

function build_linux_secured_with_rd()
{
	echo "*********** Build for linux With Resource Directory & Security ************"
	scons RELEASE=$1 WITH_RD=1 SECURED=1 $2
}

function build_android()
{
	# Note: for android, as oic-resource uses C++11 feature stoi and to_string,
	# it requires gcc-4.9, currently only android-ndk-r10(for linux)
	# and windows android-ndk-r10(64bit target version) support these features.

	build_android_x86 $1 $2
	build_android_x86_with_rm $1 $2
	build_android_armeabi $1 $2
	build_android_armeabi_with_rm $1 $2
}

function build_android_x86()
{
	echo "*********** Build for android x86 *************"
	scons TARGET_OS=android TARGET_ARCH=x86 RELEASE=$1 TARGET_TRANSPORT=IP $2
	scons TARGET_OS=android TARGET_ARCH=x86 RELEASE=$1 TARGET_TRANSPORT=BT $2
	scons TARGET_OS=android TARGET_ARCH=x86 RELEASE=$1 TARGET_TRANSPORT=BLE $2
}

function build_android_x86_with_rm()
{
	echo "*********** Build for android x86 with Routing Manager *************"
	scons TARGET_OS=android TARGET_ARCH=x86 ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=IP $2
	scons TARGET_OS=android TARGET_ARCH=x86 ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=BT $2
	scons TARGET_OS=android TARGET_ARCH=x86 ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=BLE $2
}

function build_android_armeabi()
{
	echo "*********** Build for android armeabi *************"
	scons TARGET_OS=android TARGET_ARCH=armeabi RELEASE=$1 TARGET_TRANSPORT=IP $2
	scons TARGET_OS=android TARGET_ARCH=armeabi RELEASE=$1 TARGET_TRANSPORT=BT $2
	scons TARGET_OS=android TARGET_ARCH=armeabi RELEASE=$1 TARGET_TRANSPORT=BLE $2
}

function build_android_armeabi_with_rm()
{
	echo "*********** Build for android armeabi with Routing Manager*************"
	scons TARGET_OS=android TARGET_ARCH=armeabi ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=IP $2
	scons TARGET_OS=android TARGET_ARCH=armeabi ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=BT $2
	scons TARGET_OS=android TARGET_ARCH=armeabi ROUTING=GW RELEASE=$1 TARGET_TRANSPORT=BLE $2
}

function build_arduino()
{
	echo "*********** Build for arduino avr *************"
	scons resource TARGET_OS=arduino UPLOAD=false BOARD=mega TARGET_ARCH=avr TARGET_TRANSPORT=IP SHIELD=ETH RELEASE=$1 $2
	scons resource TARGET_OS=arduino UPLOAD=false BOARD=mega TARGET_ARCH=avr TARGET_TRANSPORT=IP SHIELD=WIFI RELEASE=$1 $2
	scons resource TARGET_OS=arduino UPLOAD=false BOARD=mega TARGET_ARCH=avr TARGET_TRANSPORT=BLE SHIELD=RBL_NRF8001 RELEASE=$1 $2

	echo "*********** Build for arduino arm *************"
	scons resource TARGET_OS=arduino UPLOAD=false BOARD=arduino_due_x TARGET_ARCH=arm TARGET_TRANSPORT=IP SHIELD=ETH RELEASE=$1 $2
	scons resource TARGET_OS=arduino UPLOAD=false BOARD=arduino_due_x TARGET_ARCH=arm TARGET_TRANSPORT=IP SHIELD=WIFI RELEASE=$1 $2
	# BLE support for the Arduino Due is currently unavailable.
}

function build_tizen()
{
	echo "*********** Build for Tizen *************"
	./gbsbuild.sh

	echo "*********** Build for Tizen octbstack lib and sample *************"
	scons -f resource/csdk/stack/samples/tizen/build/SConscript TARGET_OS=tizen TARGET_TRANSPORT=IP LOGGING=true RELEASE=$1 $2

	echo "*********** Build for Tizen octbstack lib and sample with Security*************"
	scons -f resource/csdk/stack/samples/tizen/build/SConscript TARGET_OS=tizen TARGET_TRANSPORT=IP LOGGING=true SECURED=1 RELEASE=$1 $2

	echo "*********** Build for Tizen octbstack lib and sample with Routing Manager*************"
	scons -f resource/csdk/stack/samples/tizen/build/SConscript TARGET_OS=tizen TARGET_TRANSPORT=IP LOGGING=true ROUTING=GW RELEASE=$1 $2
}

function build_darwin() # Mac OSx and iOS
{
	echo "*********** Build for OSX *************"
	scons TARGET_OS=darwin SYS_VERSION=10.9 RELEASE=$1 $2

	echo "*********** Build for IOS i386 *************"
	scons TARGET_OS=ios TARGET_ARCH=i386 SYS_VERSION=7.0 RELEASE=$1 $2

	echo "*********** Build for IOS x86_64 *************"
	scons TARGET_OS=ios TARGET_ARCH=x86_64 SYS_VERSION=7.0 RELEASE=$1 $2

	echo "*********** Build for IOS armv7 *************"
	scons TARGET_OS=ios TARGET_ARCH=armv7 SYS_VERSION=7.0 RELEASE=$1 $2

	echo "*********** Build for IOS armv7s *************"
	scons TARGET_OS=ios TARGET_ARCH=armv7s SYS_VERSION=7.0 RELEASE=$1 $2

	echo "*********** Build for IOS arm64 *************"
	scons TARGET_OS=ios TARGET_ARCH=arm64 SYS_VERSION=7.0 RELEASE=$1 $2
}

function build_simulator()
{
	echo "*********** Build for simulator plugin *************"
	scons SIMULATOR=1 RELEASE=$1 $2
}

function unit_tests()
{
	echo "*********** Unit test Start *************"
	scons resource RELEASE=false -c
	scons resource LOGGING=false RELEASE=false V=1
	scons resource TEST=1 RELEASE=false V=1
	echo "*********** Unit test Stop *************"
}

function  help()
{
	echo "Usage:"
        echo "  build:"
        echo "     `basename $0` <target_build>"
	echo "      Allowed values for <target_build>: all, linux_unsecured, linux_secured, linux_unsecured_with_ra, linux_secured_with_ra, linux_unsecured_with_rd, linux_secured_with_rd, android, arduino, tizen, simulator darwin"
	echo "      Note: \"linux\" will build \"linux_unsecured\", \"linux_secured\", \"linux_unsecured_with_ra\", \"linux_secured_with_ra\", \"linux_secured_with_rd\" & \"linux_unsecured_with_rd\"."
	echo "      Any selection will build both debug and release versions of all available targets in the scope you've"
	echo "      selected. To choose any specific command, please use the SCons commandline directly. Please refer"
	echo "      to [IOTIVITY_REPO]/Readme.scons.txt."
        echo "  clean:"
        echo "     `basename $0` -c"
}

# Suppress "Reading ..." message and enable parallel build
export SCONSFLAGS="-Q -j 4"

if [ $# -eq 1 ]
then
	if [ $1 = '-c' ]
	then
		build_all true $1
		build_all false $1
		exit 0
	elif [ $1 = 'all' ]
	then
		build_all true
		build_all false
		unit_tests
	elif [ $1 = 'linux' ]
	then
		build_linux true
		build_linux false
	elif [ $1 = 'linux_unsecured' ]
	then
		build_linux_unsecured true
		build_linux_unsecured false
		build_linux_unsecured_with_rm true
		build_linux_unsecured_with_rm false
	elif [ $1 = 'linux_secured' ]
	then
		build_linux_secured true
		build_linux_secured false
	elif [ $1 = 'linux_unsecured_with_ra' ]
	then
		build_linux_unsecured_with_ra true
		build_linux_unsecured_with_ra false
	elif [ $1 = 'linux_secured_with_ra' ]
	then
		build_linux_secured_with_ra true
		build_linux_secured_with_ra false
	elif [ $1 = 'linux_unsecured_with_rd' ]
	then
		build_linux_unsecured_with_rd true
		build_linux_unsecured_with_rd false
	elif [ $1 = 'linux_secured_with_rd' ]
	then
		build_linux_secured_with_rd true
		build_linux_secured_with_rd false
	elif [ $1 = 'android' ]
	then
		build_android true
		build_android false
	elif [ $1 = 'android_x86' ]
	then
        build_android_x86 true
        build_android_x86 false
		build_android_x86_with_rm true
		build_android_x86_with_rm false
	elif [ $1 = 'android_armeabi' ]
	then
        build_android_armeabi true
        build_android_armeabi false
		build_android_armeabi_with_rm true
		build_android_armeabi_with_rm false
	elif [ $1 = 'arduino' ]
	then
		build_arduino true
		build_arduino false
	elif [ $1 = 'tizen' ]
	then
		build_tizen true
		build_tizen false
	elif [ $1 = 'simulator' ]
    then
		build_simulator true
		build_simulator false
	elif [ $1 = 'darwin' ]
	then
		build_darwin true
		build_darwin false
	elif [ $1 = 'unit_tests' ]
	then
		unit_tests
	else
		help
		exit -1
	fi
elif [ $# -eq 0 ]
then
	build_all true
	build_all false
	unit_tests
else
	help
	exit -1
fi

echo "===================== done ====================="
